### ISSUE TITLE
Consider importlib.metadata over pkg_resources for newer Python versions

### ISSUE THREAD

**Issue by dvzrv** [Kozea/Radicale#1184]
Hi! I'm packaging radicale for Arch Linux. We are currently trying to clean up packages that require setuptools during runtime (mainly those packages did so due to definition of entry_points, which are now properly handled without setuptools).

In newer Python versions it is more viable to [rely on importlib.metadata to retrieve a package version](https://docs.python.org/3.9/library/importlib.metadata.html?highlight=importlib%20metadata#distribution-versions). Would it be possible to move towards attempting to use importlib.metadata if pkg_resources is not found?

For posterity, here is an example of where pkg_resources is used in the code base (pulling in setuptools during runtime):
https://github.com/Kozea/Radicale/blob/34bec01c9bcb04cfee5d92d89baa1e8bac572a11/radicale/__init__.py#L37

**Comment by jonsger** [comment]
Hi,

I'm trying to update the GNU Guix package of radicale, when I stumbled over this issue. I got 2 of 3 uses of `pkg_resources` removed:
```diff
diff --git a/radicale/__init__.py b/radicale/__init__.py
index 1f10773..760015b 100644
--- a/radicale/__init__.py
+++ b/radicale/__init__.py
@@ -29,13 +29,13 @@ import os
 import threading
 from typing import Iterable, Optional, cast
 
-import pkg_resources
+import importlib.metadata
 
 from radicale import config, log, types
 from radicale.app import Application
 from radicale.log import logger
 
-VERSION: str = pkg_resources.get_distribution("radicale").version
+VERSION: str = importlib.metadata.version("radicale")
 
 _application_instance: Optional[Application] = None
 _application_config_path: Optional[str] = None
diff --git a/radicale/storage/__init__.py b/radicale/storage/__init__.py
index db6a871..8109b2b 100644
--- a/radicale/storage/__init__.py
+++ b/radicale/storage/__init__.py
@@ -29,7 +29,7 @@ from hashlib import sha256
 from typing import (Iterable, Iterator, Mapping, Optional, Sequence, Set,
                     Tuple, Union, overload)
 
-import pkg_resources
+import importlib.metadata
 import vobject
 
 from radicale import config
@@ -41,7 +41,7 @@ INTERNAL_TYPES: Sequence[str] = ("multifilesystem", "multifilesystem_nolock",)
 
 CACHE_DEPS: Sequence[str] = ("radicale", "vobject", "python-dateutil",)
 CACHE_VERSION: bytes = "".join(
-    "%s=%s;" % (pkg, pkg_resources.get_distribution(pkg).version)
+    "%s=%s;" % (pkg, importlib.metadata.version(pkg))
     for pkg in CACHE_DEPS).encode()
 
 
```

But https://github.com/Kozea/Radicale/blob/b64c9baa5f89929ab542e304cf0a32583fe0b500/radicale/web/internal.py#L42
seems to be a bit more tricky to replace. I tried some different approaches, but all failed. The relevant migration guide can be found here: https://importlib-resources.readthedocs.io/en/latest/migration.html#pkg-resources-resource-filename

**Comment by Unrud** [comment]
> But ... seems to be a bit more tricky to replace. I tried some different approaches, but all failed.

[importlib.resources.files](https://docs.python.org/3/library/importlib.html#importlib.resources.files) can be used to get [Traversable](https://docs.python.org/3/library/importlib.html#importlib.abc.Traversable). The function [serve_folder](https://github.com/Kozea/Radicale/blob/a97093d0010e342a0f2f2e50d8ce68e7e00f258d/radicale/httputils.py#L143) must be converted to work with Traversable instead of the file system (the Traversable can be a regular file system path but it can also point inside an compressed egg file).

The main reason for not using it, is that all required parts are only available in Python >= 3.9.

**Comment by dvzrv** [comment]
Awesome stuff! Thank you! :tada:
